p2p/src/cybershake.rs

```diff
@@ -218,73 +236,246 @@ where
     Ok((received_remote_identity, outgoing, incoming))
 }
 
-// TODO: implement AsyncWrite for this, buffering the data and encrypting on flush or on each N-byte chunk.
+macro_rules! ready {
+    ($($tokens:tt)*) => {
+        match $($tokens)* {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(Ok(n)) => { n }
+            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+        }
+    };
+}
+
 impl<W: AsyncWrite + Unpin> Outgoing<W> {
+    /// Send a message with any length.
     pub async fn send_message(&mut self, msg: &[u8]) -> Result<(), Error> {
+        let mut written = 0;
+        while written != msg.len() {
+            match self.write(&msg[written..]).await {
+                Ok(n) => written += n,
+                Err(e) => {
+                    return match e.kind() {
+                        io::ErrorKind::InvalidData => Err(Error::ProtocolError),
+                        _ => Err(Error::IoError(e)),
+                    }
+                }
+            };
+        }
+        self.flush().await.map_err(|e| Error::IoError(e))?;
+        Ok(())
+    }
+}
+
+impl<W: AsyncWrite + Unpin> Outgoing<W> {
+    fn cipher_buf(&mut self) {
+        self.flushing = true;
         self.kdf.append_u64(b"seq", self.seq);
         let mut key = [0u8; 32];
         self.kdf.challenge_bytes(b"key", &mut key);
 
         let ad = encode_u64le(self.seq);
 
         let ciphertext = Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
-            .encrypt(&[&ad], msg)
-            .map_err(|_| Error::ProtocolError)?;
+            .encrypt(&[&ad], &self.buf)
+            .map_err(|_| unimplemented!())
+            .unwrap();
+
+        self.buf.clear();
+
+        // Write length of the message.
+        Write::write(&mut self.buf, &encode_u16le(ciphertext.len() as u16)[..]).unwrap();
+        // Write a message.
+        Write::write(&mut self.buf, &ciphertext).unwrap();
 
         self.seq += 1;
+    }
 
-        // Write the length prefix and the ciphertext.
-        self.writer
-            .write(&encode_u64le(ciphertext.len() as u64)[..])
-            .await?;
-        self.writer.write(&ciphertext[..]).await?;
-        self.writer.flush().await?;
-        Ok(())
+    fn flush_pending_ciphertext(&mut self, cx: &mut Context) -> Poll<Result<(), io::Error>> {
+        loop {
+            let poll = self
```

Comment:
i'd make this method safely callable with nothing to send. So instead of `loop`, it'd be `while sent < buf.len()` and _after_ the loop we can check `if sent > 0` and then do cleanup.

Reply:
oops, and check that the flushing flag is set.

Reply:
implemented